Cannot persist Entity with several inverse one-to-many relations due to UnsupportedOperationException

### Expected Behavior

I should be possible to save the following entity with a CrudRepository:

```java
@MappedEntity
public record Entity(
  @Id @AutoPopulated Long id,
  @Relation(value = Kind.ONE_TO_MANY, cascade = Cascade.ALL, mappedBy = "entity")
  List<SubEntityA> subEntityAs,
  // a second relationship is needed to trigger the error
  @Relation(value = Kind.ONE_TO_MANY, cascade = Cascade.ALL, mappedBy = "entity")
  List<SubEntityB> subEntityBs
){};

@MappedEntity
public record SubEntityA(
  @Id @AutoPopulated Long id,
  @Nullable Integer data,
  // a back reference is needed to trigger the error
  @Relation(Kind.MANY_TO_ONE)
  Entity entity
){};

@MappedEntity
public record SubEntityA(
  @Id @AutoPopulated Long id,
  @Nullable Integer data,
  @Relation(Kind.MANY_TO_ONE)
  Entity entity
){};

interface EntityRepository extends CrudRepository<Entity, Long> {}
```

### Actual Behaviour

When saving with `CrudRepository.save(entity)` the following error is thrown:

```
io.micronaut.data.exceptions.DataAccessException: Error executing PERSIST: null

  at io.micronaut.data.runtime.operations.internal.BaseOperations.failed(BaseOperations.java:141)
  at io.micronaut.data.runtime.operations.internal.BaseOperations.persist(BaseOperations.java:90)
  at io.micronaut.data.runtime.operations.internal.EntityOperations.persist(EntityOperations.java:31)
  at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$persist$20(DefaultJdbcRepositoryOperations.java:706)
  at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.lambda$executeWrite$24(DefaultJdbcRepositoryOperations.java:781)
  at io.micronaut.data.connection.support.AbstractConnectionOperations.withExistingConnectionInternal(AbstractConnectionOperations.java:128)
  at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:92)
  at io.micronaut.data.connection.ConnectionOperations.executeWrite(ConnectionOperations.java:82)
  at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.executeWrite(DefaultJdbcRepositoryOperations.java:778)
  at io.micronaut.data.jdbc.operations.DefaultJdbcRepositoryOperations.persist(DefaultJdbcRepositoryOperations.java:702)
  at io.micronaut.data.runtime.intercept.DefaultSaveEntityInterceptor.intercept(DefaultSaveEntityInterceptor.java:45)
  at io.micronaut.data.runtime.intercept.DataIntroductionAdvice.intercept(DataIntroductionAdvice.java:83)
  at io.micronaut.aop.chain.MethodInterceptorChain.proceed(MethodInterceptorChain.java:138)
  at mytest.EntityRepository$Intercepted.save(Unknown Source)
  at mytest.EntityRepositoryIT.testCascadingSaveAndFetching(EntityRepositoryIT.java:21)
  at java.base/java.lang.reflect.Method.invoke(Method.java:568)
  at io.micronaut.test.extensions.junit5.MicronautJunit5Extension$2.proceed(MicronautJunit5Extension.java:142)
  at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:162)
  at io.micronaut.test.extensions.AbstractMicronautExtension$3.proceed(AbstractMicronautExtension.java:174)
  at io.micronaut.test.context.TestMethodInterceptor.interceptTest(TestMethodInterceptor.java:46)
  at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.lambda$interceptTest$0(DefaultTestTransactionExecutionListener.java:93)
  at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.lambda$execute$2(AbstractPropagatedStatusTransactionOperations.java:68)
  at io.micronaut.transaction.TransactionCallback.apply(TransactionCallback.java:37)
  at io.micronaut.transaction.support.AbstractTransactionOperations.executeTransactional(AbstractTransactionOperations.java:333)
  at io.micronaut.transaction.support.AbstractTransactionOperations.executeWithNewTransaction(AbstractTransactionOperations.java:318)
  at io.micronaut.transaction.support.AbstractTransactionOperations.executeNew(AbstractTransactionOperations.java:235)
  at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:137)
  at io.micronaut.transaction.support.AbstractTransactionOperations.lambda$doExecute$0(AbstractTransactionOperations.java:122)
  at io.micronaut.data.connection.support.AbstractConnectionOperations.executeWithNewConnection(AbstractConnectionOperations.java:143)
  at io.micronaut.data.connection.support.AbstractConnectionOperations.execute(AbstractConnectionOperations.java:90)
  at io.micronaut.transaction.support.AbstractTransactionOperations.doExecute(AbstractTransactionOperations.java:120)
  at io.micronaut.transaction.support.AbstractPropagatedStatusTransactionOperations.execute(AbstractPropagatedStatusTransactionOperations.java:65)
  at io.micronaut.transaction.test.DefaultTestTransactionExecutionListener.interceptTest(DefaultTestTransactionExecutionListener.java:91)
  at io.micronaut.test.extensions.AbstractMicronautExtension.interceptEach(AbstractMicronautExtension.java:166)
  at io.micronaut.test.extensions.AbstractMicronautExtension.interceptTest(AbstractMicronautExtension.java:119)
  at io.micronaut.test.extensions.junit5.MicronautJunit5Extension.interceptTestMethod(MicronautJunit5Extension.java:129)
  at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
  at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
Caused by: java.lang.UnsupportedOperationException
  at java.base/java.util.ImmutableCollections.uoe(ImmutableCollections.java:142)
  at java.base/java.util.ImmutableCollections$ListItr.set(ImmutableCollections.java:419)
  at io.micronaut.data.runtime.operations.internal.AbstractCascadeOperations.afterCascadedMany(AbstractCascadeOperations.java:184)
  at io.micronaut.data.runtime.operations.internal.SyncCascadeOperations.cascadeEntity(SyncCascadeOperations.java:157)
  at io.micronaut.data.runtime.operations.internal.AbstractSyncEntityOperations.cascadePost(AbstractSyncEntityOperations.java:83)
  at io.micronaut.data.runtime.operations.internal.BaseOperations.persist(BaseOperations.java:87)
  ... 36 more
```

The reason seems to be `io.micronaut.data.runtime.operations.internal.AbstractSyncEntitiesOperations#getEntities` which is returning an immutable list of entities. But in `io.micronaut.data.runtime.operations.internal.AbstractCascadeOperations#afterCascadedMany` an iterator is created for this list of entities and `iterator.set(newc)` is called. For this to work we need an iterator over a mutable list.

This is possibly a regression bug due to this PR: https://github.com/micronaut-projects/micronaut-data/pull/2905


### Steps To Reproduce

1. see code sample above for entity mapping
2. create an entity with subentity A and subentity B set and save it to the repository.


### Environment Information

- Error was observed in `io.micronaut.data:micronaut-data-runtime:4.8.1` after an upgrade to `micronaut-parent` 4.5.0
- When downgrading to Micronaut 4.4.3 which is using `io.micronaut.data:micronaut-data-runtime:4.7.1` the code behaves as expected.

### Example Application

_No response_

### Version

4.5.0

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2CascadeSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2CascadeSpec.groovy
@@ -1,0 +1,41 @@
+package io.micronaut.data.jdbc.h2
+
+import io.micronaut.context.annotation.Property
+import io.micronaut.core.annotation.Nullable
+import io.micronaut.data.annotation.GeneratedValue
+import io.micronaut.data.annotation.Id
+import io.micronaut.data.annotation.Join
+import io.micronaut.data.annotation.MappedEntity
+import io.micronaut.data.annotation.Relation
+import io.micronaut.data.jdbc.annotation.JdbcRepository
+import io.micronaut.data.model.query.builder.sql.Dialect
+import io.micronaut.data.repository.CrudRepository
+import io.micronaut.test.extensions.spock.annotation.MicronautTest
+import jakarta.inject.Inject
+import spock.lang.Shared
+import spock.lang.Specification
+
+
+@MicronautTest
+@H2DBProperties
+@Property(name = "datasources.default.packages", value = "io.micronaut.data.jdbc.h2")
+class H2CascadeSpec extends Specification {
+
+    @Inject
+    @Shared
+    CascadeEntityRepository repository
+
+    void "test cascade save"() {
+        when:
+        def entityA = new CascadeSubEntityA(null, 1, null)
+        def entityB = new CascadeSubEntityB(null, 2, null)
+        def entity = new CascadeEntity(null, List.of(entityA), List.of(entityB))
+        entity = repository.save(entity)
+        def opt = repository.findById(entity.id)
+        then:
+        opt.present
+        def loadedEntity = opt.get()
+        loadedEntity.subEntityAs.size() == 1
+        loadedEntity.subEntityBs.size() == 1
+    }
+}

--- a/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2CascadeSpec.groovy
+++ b/data-jdbc/src/test/groovy/io/micronaut/data/jdbc/h2/H2CascadeSpec.groovy
@@ -1,15 +1,6 @@
 package io.micronaut.data.jdbc.h2
 
 import io.micronaut.context.annotation.Property
-import io.micronaut.core.annotation.Nullable
-import io.micronaut.data.annotation.GeneratedValue
-import io.micronaut.data.annotation.Id
-import io.micronaut.data.annotation.Join
-import io.micronaut.data.annotation.MappedEntity
-import io.micronaut.data.annotation.Relation
-import io.micronaut.data.jdbc.annotation.JdbcRepository
-import io.micronaut.data.model.query.builder.sql.Dialect
-import io.micronaut.data.repository.CrudRepository
 import io.micronaut.test.extensions.spock.annotation.MicronautTest
 import jakarta.inject.Inject
 import spock.lang.Shared
@@ -31,11 +22,11 @@ class H2CascadeSpec extends Specification {
         def entityB = new CascadeSubEntityB(null, 2, null)
         def entity = new CascadeEntity(null, List.of(entityA), List.of(entityB))
         entity = repository.save(entity)
-        def opt = repository.findById(entity.id)
+        def opt = repository.findById(entity.id())
         then:
         opt.present
         def loadedEntity = opt.get()
-        loadedEntity.subEntityAs.size() == 1
-        loadedEntity.subEntityBs.size() == 1
+        loadedEntity.subEntityAs().size() == 1
+        loadedEntity.subEntityBs().size() == 1
     }
 }

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
@@ -12,7 +12,6 @@ public record CascadeEntity(
         @Id @GeneratedValue Long id,
         @Relation(value = Relation.Kind.ONE_TO_MANY, cascade = Relation.Cascade.ALL, mappedBy = "entity")
         List<CascadeSubEntityA> subEntityAs,
-        // a second relationship is needed to trigger the error
         @Relation(value = Relation.Kind.ONE_TO_MANY, cascade = Relation.Cascade.ALL, mappedBy = "entity")
         List<CascadeSubEntityB> subEntityBs
 ){};

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
@@ -14,4 +14,4 @@ public record CascadeEntity(
         List<CascadeSubEntityA> subEntityAs,
         @Relation(value = Relation.Kind.ONE_TO_MANY, cascade = Relation.Cascade.ALL, mappedBy = "entity")
         List<CascadeSubEntityB> subEntityBs
-){};
+){}

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntity.java
@@ -1,0 +1,18 @@
+package io.micronaut.data.jdbc.h2;
+
+import io.micronaut.data.annotation.GeneratedValue;
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+import io.micronaut.data.annotation.Relation;
+
+import java.util.List;
+
+@MappedEntity
+public record CascadeEntity(
+        @Id @GeneratedValue Long id,
+        @Relation(value = Relation.Kind.ONE_TO_MANY, cascade = Relation.Cascade.ALL, mappedBy = "entity")
+        List<CascadeSubEntityA> subEntityAs,
+        // a second relationship is needed to trigger the error
+        @Relation(value = Relation.Kind.ONE_TO_MANY, cascade = Relation.Cascade.ALL, mappedBy = "entity")
+        List<CascadeSubEntityB> subEntityBs
+){};

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntityRepository.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeEntityRepository.java
@@ -1,0 +1,11 @@
+package io.micronaut.data.jdbc.h2;
+
+import io.micronaut.data.annotation.Join;
+import io.micronaut.data.jdbc.annotation.JdbcRepository;
+import io.micronaut.data.model.query.builder.sql.Dialect;
+import io.micronaut.data.repository.CrudRepository;
+
+@JdbcRepository(dialect = Dialect.H2)
+@Join("subEntityAs")
+@Join("subEntityBs")
+interface CascadeEntityRepository extends CrudRepository<CascadeEntity, Long> {}

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityA.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityA.java
@@ -1,0 +1,16 @@
+package io.micronaut.data.jdbc.h2;
+
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.data.annotation.GeneratedValue;
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+import io.micronaut.data.annotation.Relation;
+
+@MappedEntity
+public record CascadeSubEntityA(
+        @Id @GeneratedValue Long id,
+        @Nullable Integer data,
+        @Relation(Relation.Kind.MANY_TO_ONE)
+        @Nullable
+        CascadeEntity entity
+){};

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityA.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityA.java
@@ -13,4 +13,4 @@ public record CascadeSubEntityA(
         @Relation(Relation.Kind.MANY_TO_ONE)
         @Nullable
         CascadeEntity entity
-){};
+){}

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityB.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityB.java
@@ -1,0 +1,16 @@
+package io.micronaut.data.jdbc.h2;
+
+import io.micronaut.core.annotation.Nullable;
+import io.micronaut.data.annotation.GeneratedValue;
+import io.micronaut.data.annotation.Id;
+import io.micronaut.data.annotation.MappedEntity;
+import io.micronaut.data.annotation.Relation;
+
+@MappedEntity
+public record CascadeSubEntityB(
+        @Id @GeneratedValue Long id,
+        @Nullable Integer data,
+        @Relation(Relation.Kind.MANY_TO_ONE)
+        @Nullable
+        CascadeEntity entity
+){};

--- a/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityB.java
+++ b/data-jdbc/src/test/java/io/micronaut/data/jdbc/h2/CascadeSubEntityB.java
@@ -13,4 +13,4 @@ public record CascadeSubEntityB(
         @Relation(Relation.Kind.MANY_TO_ONE)
         @Nullable
         CascadeEntity entity
-){};
+){}

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/AbstractSyncEntitiesOperations.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/AbstractSyncEntitiesOperations.java
@@ -32,6 +32,7 @@ import java.util.Objects;
 import java.util.function.Consumer;
 import java.util.function.Function;
 import java.util.function.Predicate;
+import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 /**
@@ -157,7 +158,7 @@ public abstract class AbstractSyncEntitiesOperations<Ctx extends OperationContex
 
     @Override
     public List<T> getEntities() {
-        return entities.stream().map(d -> d.entity).toList();
+        return entities.stream().map(d -> d.entity).collect(Collectors.toList());
     }
 
     @SuppressWarnings("VisibilityModifier")

--- a/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/AbstractSyncEntitiesOperations.java
+++ b/data-runtime/src/main/java/io/micronaut/data/runtime/operations/internal/AbstractSyncEntitiesOperations.java
@@ -157,6 +157,7 @@ public abstract class AbstractSyncEntitiesOperations<Ctx extends OperationContex
     }
 
     @Override
+    @SuppressWarnings({"java:S6204"})
     public List<T> getEntities() {
         return entities.stream().map(d -> d.entity).collect(Collectors.toList());
     }

